engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,575 @@ package ingestion
+		// if one stop was set by the version beacon and the other one was manual
+		// we can only update if the new stop is earlier
+
+		// this prevents users moving the stopHeight forward when a version boundary
+		// is earlier, and prevents version beacons from moving the stopHeight forward
+		// when a manual stop is earlier.
+		return false, "cannot update stopHeight, new stop height is later than the " +
+			"current one (or removing a stop)"
+	}
 
-	return oldHeight, oldCrash, nil
+	return true, ""
 }
 
-// GetStopHeight returns:
-//   - stopHeight
-//   - crash
-//
-// Values are undefined if they were not previously set
-func (s *StopControl) GetStopHeight() (uint64, bool) {
+// GetStop returns the upcoming stop parameters or nil if no stop is set.
+func (s *StopControl) GetStop() *StopParameters {
 	s.RLock()
 	defer s.RUnlock()
 
-	return s.stopHeight, s.crash
+	if s.stopBoundary == nil {
+		return nil
+	}
+
+	p := s.stopBoundary.StopParameters
+	return &p
 }
 
-// blockProcessable should be called when new block is processable.
+// BlockProcessable should be called when new block is processable.
 // It returns boolean indicating if the block should be processed.
-func (s *StopControl) blockProcessable(b *flow.Header) bool {
+func (s *StopControl) BlockProcessable(b *flow.Header) bool {
 	s.Lock()
 	defer s.Unlock()
 
-	if s.state == StopControlOff {
-		return true
+	// don't process anymore blocks if stopped
+	if s.stopped {
+		return false
 	}
 
-	if s.state == StopControlPaused {
-		return false
+	// if no stop is set process all blocks
+	if s.stopBoundary == nil {
+		return true
 	}
 
-	// skips blocks at or above requested stopHeight
-	if b.Height >= s.stopHeight {
+	// Skips blocks at or above requested stopHeight
+	// doing so means we have started the stopping process
+	if b.Height >= s.stopBoundary.StopHeight {
 		s.log.Warn().
-			Int8("previous_state", int8(s.state)).
-			Int8("new_state", int8(StopControlCommenced)).
 			Msgf(
 				"Skipping execution of %s at height %d"+
-					" because stop has been requested at height %d",
+					" because stop has been requested %s",
 				b.ID(),
 				b.Height,
-				s.stopHeight,
+				s.stopBoundary,
 			)
 
-		s.state = StopControlCommenced // if block was skipped, move into commenced state
+		s.stopBoundary.cannotBeChanged = true
 		return false
 	}
 
 	return true
 }
 
-// blockFinalized should be called when a block is marked as finalized
-func (s *StopControl) blockFinalized(
+// BlockFinalized should be called when a block is marked as finalized
+//
+// Once finalization reached stopHeight we can be sure no other fork will be valid at
+// this height, if this block's parent has been executed, we are safe to stop.
+// This will happen during normal execution, where blocks are executed
+// before they are finalized. However, it is possible that EN block computation
+// progress can fall behind. In this case, we want to crash only after the execution
+// reached the stopHeight.
```

Comment:
This description makes me to think of another idea to possibly simply the logic here.

The fact the we need to handle BlockFinalized event is because an executed block might not be finalized. So if we stop on a block that has not been finalized even if it has reached the stop height, then there is a chance that block happens to be a unfinalized block, and the real finalized block will never be executed, because we have stopped.

We are trying to distinguish the following two cases, and handle them differently:
1. execution is ahead of finalization
2. finalization is ahead of execution

My idea is that we can build a module (FinalizedExecuted) that takes `OnBlockFinalized` and `OnBlockExecuted` events, and returns the latest `FinalizedAndExecutedBlockHeight`. So that we can simplify the stopBoundary by removing the `stopAfterExecuting` block ID field. And move it to the FinalizedExecuted module. 

Then the complexity is break down into two problems:
1. When a block is finalized or executed, calculate what's the highest finalized and executed height
2. Check if the highest finalized and executed height has reached the stop height. 

And it will also be easier to debug, because if it should stop but didn't stop, we could first check the log and see if the highest finalized and executed height was determined correctly or not, which gives us a point to break the large debugging surface into half.